linkerd/proxy/server-policy/src/local_rate_limit.rs

```diff
@@ -0,0 +1,173 @@
+#[cfg(test)]
+use governor::clock::FakeRelativeClock;
+use governor::{
+    clock::{Clock, DefaultClock},
+    middleware::NoOpMiddleware,
+    state::{keyed::HashMapStateStore, InMemoryState, RateLimiter, StateStore},
+    Quota,
+};
+use linkerd_identity::Id;
+use std::{collections::HashMap, num::NonZeroU32, sync::Arc};
+
+#[cfg(test)]
+mod tests;
+
+type Direct = InMemoryState;
+type Keyed = HashMapStateStore<Option<Id>>;
+
+#[derive(Debug, Default)]
+pub struct LocalRateLimit<C: Clock = DefaultClock> {
+    total: Option<RateLimit<Direct, C>>,
+    per_identity: Option<RateLimit<Keyed, C>>,
+    overrides: HashMap<Id, Arc<RateLimit<Direct, C>>>,
+}
+
+#[derive(Debug)]
+struct RateLimit<S, C = DefaultClock>
+where
+    S: StateStore,
+    C: Clock,
+{
+    rps: NonZeroU32,
+    limiter: RateLimiter<S::Key, S, C, NoOpMiddleware<C::Instant>>,
+}
+
+#[derive(Debug, thiserror::Error, PartialEq, Eq)]
+pub enum RateLimitError {
+    #[error("total rate limit exceeded: {0}rps")]
+    Total(NonZeroU32),
+    #[error("per-identity rate limit exceeded: {0}rps")]
+    PerIdentity(NonZeroU32),
+    #[error("override rate limit exceeded: {0}rps")]
+    Override(NonZeroU32),
+}
+
+// === impl LocalRateLimit ===
+
+impl LocalRateLimit {
+    pub fn new_no_overrides(
```

Comment:
We would typically [feature-gate this](https://github.com/linkerd/linkerd2-proxy/pull/3328) to assert that it's only used for tests.